conjure-java-core/src/main/java/com/palantir/conjure/java/GeneratedFile.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java;
+
+import com.palantir.conjure.java.GeneratedFile.GeneratedJavaFile;
+import com.palantir.conjure.java.GeneratedFile.GeneratedReachabilityMetadataFile;
+import com.palantir.conjure.java.types.ReachabilityMetadata;
+import com.palantir.javapoet.JavaFile;
+
+public sealed interface GeneratedFile permits GeneratedJavaFile, GeneratedReachabilityMetadataFile {
+    record GeneratedJavaFile(JavaFile javaFile) implements GeneratedFile {
+        public static GeneratedJavaFile of(JavaFile javaFile) {
+            return new GeneratedJavaFile(javaFile);
+        }
+    }
+
+    record GeneratedReachabilityMetadataFile(String packageName, ReachabilityMetadata reachabilityMetadata)
```

Comment:
Generating one JSON per package here. An alternative is to create one JSON per type. This is what we'd want to do when generating metadata in an annotation library as it allows Gradle to make use of incremental compilation: skip generation of metadata for classes that have not changed.

For `conjure-java`, I can go either way. One JSON per type is a load of files and slightly more difficult to inspect manually so I opted to do this. (it's not really all that much more difficult - I can just `cat` all the files in a directory)